### PR TITLE
[FLINK-3749] [table] Improve decimal handling

### PR DESCRIPTION
**Pull request by twalthr**
This PR improves the decimal handling and fixes several bugs.

### CODE REVIEW

File: flink-libraries/flink-table/src/main/scala/org/apache/flink/api/table/codegen/CodeGenerator.scala [hunk 1]
```diff
@@ -551,23 +551,27 @@ class CodeGenerator(
...
         else {
           throw new CodeGenException("Decimal can not be converted to float.")
         }
-      case DOUBLE =>
+      case DOUBLE | DECIMAL =>
         val decimal = BigDecimal(value.asInstanceOf[java.math.BigDecimal])
-        if (decimal.isValidDouble) {
-          generateNonNullLiteral(resultType, decimal.doubleValue().toString)
+        // check if we loose/change digits when converting to double
+        val converted = BigDecimal(decimal.doubleValue().toString)
```
Comment (ramkrish86) [inline flink-libraries/flink-table/src/main/scala/org/apache/flink/api/table/codegen/CodeGenerator.scala]: So with this change all the 11.CAST(STRING) and 11.abs() all should work right?

Comment (twalthr) [inline flink-libraries/flink-table/src/main/scala/org/apache/flink/api/table/codegen/CodeGenerator.scala]: No, thats a different issue. The `ExpressionParser` needs a large rework for all that.

**Comment by fhueske** [comment]
I don't think we should handle `DECIMAL` as `double`. `DECIMAL` provides fixed precision in contrast to `double`. If we internally represent `DECIMAL` as `double` precision is not guaranteed. I think we need a special data type, serializers, and also custom math logic to handle `DECIMAL` correctly. Maybe Calcite offers some of that?

**Comment by twalthr** [comment]
Yes, you are right. We need special data type etc. for that in the future. But until then we should treat it as double as long as we can represent it. I can create an issue for that.

**Comment by fhueske** [comment]
> There is nothing more permanent than than a temporary solution ;-)

Please correct me, if I got something wrong.
This change adds support for `DECIMAL` literals. So we might have row fields of `DECIMAL` type and  might compute incorrect results until we add proper support for `DECIMAL`. 

It would be nice to support `DECIMAL` but I don't think it is more important than result correctness. Depending on the support Calcite (or other libraries) provide, adding proper `DECIMAL` support might be a complex task and might take some time.

Is there an urgent need to support `DECIMAL`?

**Comment by twalthr** [comment]
> It's better to have a temporary solution than no solution ;-)

Calcite translates all decimal literals to `DECIMAL`. At the moment even `SELECT 11.2 FROM MyTable` does not work, although it could fit into `DOUBLE`. I thought it makes sense to support at least decimals of `DOUBLE` precision.

**Comment by vasia** [comment]
@twalthr I see your point. Is it possible to support at least decimals of double precision without running into the risk of returning wrong results? If not, I don't think we should go for this temporary solution.

**Comment by twalthr** [comment]
@vasia and @fhueske: I have opened FLINK-3786 for a long-term solution ;-)
I will start working on this issue.

**Comment by fhueske** [comment]
Thanks @twalthr ! 
Do you think solving FLINK-3786 will be a lot of effort?

I did not know that Calcite internally handles all fraction literals as DECIMAL. That explains the need for this data type.

**Comment by twalthr** [comment]
Solving FLINK-3786 should not be too complicated. Integrating it into the Table API might need some time.

**Comment by fhueske** [comment]
@twalthr, can this PR be closed in favor of #2088?

**Comment by twalthr** [comment]
Yes, I will close this...

**Pull request closed without merging**